### PR TITLE
[stdlib] Add consuming/owned annotations to Collection implementations

### PR DESCRIPTION
**Pull request by airspeedswift**
Mostly following the protocols, various ownership annotations on collection types:

- make all (non-trivial) iterators `__consuming`
- all appended/inserted elements `__owned`
- make casting-like operations `__consuming`

Most changes are in now. I'm leaving `Set` and `Dictionary` for later as they're in-flux elsewhere.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please smoke test

**Comment by airspeedswift** [comment]
@swift-ci please test

File: test/IDE/reconstruct_type_from_mangled_name.swift [hunk 1]
```diff
@@ -64,22 +64,22 @@ class Myclass2 {
 
     arr1.append(1)
 // FIXME: missing append()
-// CHECK: dref: FAILURE	for 'append' usr=s:Sa6appendyyxF
-// CHECK: type: (inout Array<Int>) -> (Int) -> ()
```
Comment (airspeedswift) [inline test/IDE/reconstruct_type_from_mangled_name.swift]: @nkcsgexi @benlangmuir not entirely sure what this was testing so the fix here was just to make it match the new output... `Array.append(_:)` now takes its argument owned so the mangled name changed but not sure about the following line.

File: test/SILOptimizer/array_element_propagation.sil [hunk 2]
```diff
@@ -357,13 +357,12 @@ sil [_semantics "array.append_contentsOf"] @arrayAppendContentsOfHello : $@conve
...
 // CHECK-NEXT: store %1 to %{{[0-9]+}} : $*Hello
 // CHECK-NOT:     apply
-// CHECK:      [[AEFUN:%.*]] = function_ref @$sSa6appendyyxF
+// CHECK:      [[AEFUN:%.*]] = function_ref @$sSa6appendyyxnF
 // CHECK-NEXT: strong_retain %1 : $Hello
 // CHECK-NEXT: [[STACK:%.*]] = alloc_stack $Hello
 // CHECK-NEXT: store %1 to [[STACK]]
 // CHECK-NEXT: apply [[AEFUN]]<Hello>([[STACK]], %0)
 // CHECK-NEXT: dealloc_stack [[STACK]]
-// CHECK-NEXT: strong_release %1
```
Comment (airspeedswift) [inline test/SILOptimizer/array_element_propagation.sil]: @eeckstein this was the only non-mangled-name change that was needed

File: test/IDE/reconstruct_type_from_mangled_name.swift (hunk 1 above)
Comment (benlangmuir) [reply test/IDE/reconstruct_type_from_mangled_name.swift]: The type change from `(inout Array<Int>) -> (Int) -> ()` to `(inout Array<Int>) -> () -> ()` is bogus.  I assume that means we aren't handling owned et. al. correctly in type reconstruction. @slavapestov what do you think?

**Comment by airspeedswift** [comment]
Oops missed an assert-build-only optimizer test...

File: test/IDE/reconstruct_type_from_mangled_name.swift (hunk 1 above)
Comment (airspeedswift) [reply test/IDE/reconstruct_type_from_mangled_name.swift]: So would it be better to XFAIL this one for now?

**Comment by airspeedswift** [comment]
@swift-ci please test

File: test/IDE/reconstruct_type_from_mangled_name.swift (hunk 1 above)
Comment (slavapestov) [reply test/IDE/reconstruct_type_from_mangled_name.swift]: You can go ahead and XFAIL it, please do file a radar though.

**Comment by airspeedswift** [comment]
@swift-ci please smoke test

**Review by eeckstein — APPROVED** [review]
lgtm

**Pull request merged**